text/0000-argument-lifetimes.md

```diff
@@ -0,0 +1,476 @@
+    fn bar(&self, arg: &'b str) -> &'b str;
+}
+
+fn take_fn(x: &u32, y: fn(&'x u32, &'b u32, &'b u32));
+```
+
+## The wildcard lifetime
+
+When referring to a type (other than `&`/`&mut`) that requires lifetime
+arguments, it is deprecated to leave off those parameters.
+
+Instead, you can write a `'_` for the parameters, rather than giving a lifetime
+name, which will have identical behavior to leaving them off today.
+
+Thus, where today you would write:
+
+```rust
+fn foo(&self) -> Ref<SomeType>
+fn iter(&self) -> Iter<T>
+```
+
+tomorrow you would write:
+
+```rust
+fn foo(&self) -> Ref<'_, SomeType>
+fn iter(&self) -> Iter<'_, T>
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The deprecations here involve some amount of churn (in the form of deleting
+lifetimes from `<>` blocks) if not ignored. Users exercise a lot of control over
+when they address that churn and can do so incrementally, and we can likely
+provide an automated tool for switching to the new style.
+
+The fact that lifetime parameters are not bound in an out-of-band way is
+somewhat unusual and might be confusing---but then, so are lifetime parameters!
+Putting the bindings out of band buys us very little, as argued in the next
+section.
+
+Requiring a `'_` rather than being able to leave off lifetimes altogether may be
+a slight decrease in ergonomics in some cases. In particular, `SomeType<'_>` is
+pretty sigil-heavy.
+
+Cases where you could write `fn foo<'a, 'b: 'a>(...)` now need the `'b: 'a` to
+be given in a `where` clause, which might be slightly more verbose. These are
+relatively rare, though, due to our type well-formedness rule.
+
+Otherwise, it's a bit hard to see drawbacks here: nothings is made more explicit
+or harder to determine, since the binding structure continues to be completely
+unambiguous; ergonomics and, arguably, learnability both improve. And
+signatures become less noisy and easier to read.
+
+# Rationale and Alternatives
+[alternatives]: #alternatives
+
+## Core rationale
+
+The key insight of the proposed design is that out-of-band bindings for lifetime
+parameters is buying us very little today:
+
+- For free functions, it's completely unnecessary; the only lifetime "in scope"
+  is `'static`, so everything else *must* be a parameter.
+- For functions within `impl` blocks, it is solely serving the purpose of
+  distinguishing between lifetimes bound by the `impl` header and those bounds
+  by the `fn`.
+
+While this might change if we ever allow modules to be parameterized by
+lifetimes, it won't change in any essential way: the point is that there are
+generally going to be *very* few in-scope lifetimes when writing a function
+signature. We can likely use conventions or some other mechanism to help
+distinguish between the `impl` header and `fn` bindings, if needed.
+
+This RFC proposes to impose a strict distinction between lifetimes introduced in
+`impl` headers and `fn` signatures. We could instead impose a distinction
```

Comment:
I think this text refers to something that was part of the original draft but removed later?